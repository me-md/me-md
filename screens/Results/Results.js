import React, { useState, useEffect } from 'react';
import {
  Dimensions,
  StyleSheet,
  SafeAreaView,
  View,
  FlatList
} from 'react-native';
import {
  Accordion,
  Body,
  Button,
  Card,
  CardItem,
  Icon,
  Text
} from 'native-base';
import * as Progress from 'react-native-progress';
import { Header } from '../../components/Header';
import { specifyTargetCondition } from '../../utils/helpers/helpers';
import { ScrollView } from 'react-native-gesture-handler';
import { getExplanation } from '../../utils/apiCalls/Infermedica/getExplanation';
import { getConditionById } from '../../utils/apiCalls/Conditions/getConditionById';

const height = Dimensions.get('window').height;

export default function SymptomsQA({ navigation }) {
  let {
    userInfo,
    symptomFollowup,
    location,
    stateAbbreviation
  } = navigation.state.params;

  const [explanation, setExplanation] = useState({});
  const [conditionDetails, setConditionDetails] = useState({});
  const [expanded, setExpanded] = useState({});

  useEffect(() => {
    getResults();
    console.log(explanation);
  }, []);

  const getResults = async () => {
    try {
      userInfo = specifyTargetCondition(userInfo, symptomFollowup);
      let explanation = await getExplanation(userInfo);
      let conditionDetails = await getConditionById(
        symptomFollowup.conditions[0].id
      );
      setExplanation(explanation);
      setConditionDetails(conditionDetails);
    } catch (error) {
      throw new Error(error);
    }
  };

  const renderHeader = (item, expanded) => {
    return (
      <View
        style={{
          flexDirection: 'row',
          padding: 10,
          justifyContent: 'space-between',
          alignItems: 'center',
          backgroundColor: '#0960FF',
          width: '100%'
        }}
      >
        <Text style={{ fontWeight: '600', color: '#fff' }}> {item.title}</Text>
        {expanded ? (
          <Icon style={{ fontSize: 18, color: '#fff' }} name='arrow-up' />
        ) : (
          <Icon style={{ fontSize: 18, color: '#fff' }} name='arrow-down' />
        )}
      </View>
    );
  };
  const renderSupportingContent = item => {
    return (
      <SafeAreaView
        style={{
          flex: 1,
          height: expanded ? null : 0,
          overflow: 'hidden'
        }}
      >
        <FlatList
          data={createSupportingEvidence()}
          style={{ padding: height * 0.005 }}
          renderItem={({ item }) => <Text style={styles.item}>{item.key}</Text>}
        />
      </SafeAreaView>
    );
  };

  const renderConflictingContent = item => {
    return explanation.conflicting_evidence.length ? (
      <SafeAreaView
        style={{
          flex: 1,
          height: expanded ? null : 0,
          overflow: 'hidden'
        }}
      >
        <FlatList
          data={createConflictingEvidence()}
          style={{ padding: height * 0.005 }}
          renderItem={({ item }) => <Text style={styles.item}>{item.key}</Text>}
        />
      </SafeAreaView>
    ) : (
      <Text>No conflicting evidence</Text>
    );
  };

  const createTopDiagnosis = () => {
    return (
      <Card style={styles.topDiagnosis}>
        <CardItem>
          <Body>
            <Text
              style={styles.conditionName}
            >{`${conditionDetails.data.name} (${conditionDetails.data.common_name})`}</Text>
            <Progress.Bar
              style={{ marginBottom: height * 0.01 }}
              progress={symptomFollowup.conditions[0].probability}
              width={200}
            />
            <Text style={styles.probability}>{`Probablity: ${(
              symptomFollowup.conditions[0].probability * 100
            ).toFixed(0)}%`}</Text>
            <Text
              style={styles.recommendation}
            >{`Recommendation: ${conditionDetails.data.triage_level}. ${conditionDetails.data.hint}`}</Text>
            <View style={styles.btnTextHolder}>
              <Accordion
                dataArray={[
                  {
                    title: 'Supporting Evidence'
                  }
                ]}
                renderHeader={renderHeader}
                renderContent={renderSupportingContent}
                style={styles.accordion}
              />
              <Accordion
                dataArray={[
                  {
                    title: 'Conflicting Evidence'
                  }
                ]}
                renderHeader={renderHeader}
                renderContent={renderConflictingContent}
                style={styles.accordion}
              />
            </View>
          </Body>
        </CardItem>
      </Card>
    );
  };

  const createSupportingEvidence = () => {
    return explanation.supporting_evidence.map((evidence, index) => {
      return { key: `• ${evidence.common_name}` };
    });
  };

  const createConflictingEvidence = () => {
    return explanation.conflicting_evidence.map((evidence, index) => {
      return { key: `• ${evidence.common_name}` };
    });
  };

  const getTopDiagnoses = () => {
    return symptomFollowup.conditions.slice(1, 4).map((condition, index) => {
      return (
        <Card key={index} style={styles.conditionCard}>
          <CardItem>
            <Body>
              <Text
                style={styles.secondaryCondition}
              >{`${condition.name} (${condition.common_name})`}</Text>
              <Progress.Bar
                style={{ marginBottom: height * 0.01 }}
                progress={condition.probability}
                width={200}
              />
              <Text>{`Probability: ${(condition.probability * 100).toFixed(
                0
              )}%`}</Text>
            </Body>
          </CardItem>
        </Card>
      );
    });
  };

  return (
    <View style={styles.container}>
      <Header />
      <ScrollView style={styles.contentContainer}>
        {conditionDetails.data ? createTopDiagnosis() : null}
        {getTopDiagnoses()}
      </ScrollView>
      <Button
        block
        style={styles.button}
        onPress={() =>
          navigation.push('Doctors', {
            location,
            stateAbbreviation,
            userInfo,
            symptomFollowup,
            conditionDetails,
            explanation
          })
        }
      >
        <Text style={styles.buttonText}>FIND DOCTORS</Text>
      </Button>
    </View>
  );
}

const styles = StyleSheet.create({
  gradientBackground: {
    alignItems: 'center',
    flex: 1
  },
  container: {
    flex: 1,
    width: '100%'
  },
  contentContainer: {
    flex: 1,
    marginBottom: height * 0.05,
    marginTop: height * 0.02,
    width: '100%'
  },
  conditionName: {
    fontSize: 20,
    fontWeight: 'bold',
    marginBottom: height * 0.01
  },
  recommendation: {
    marginTop: height * 0.01
  },
  symptomText: {
    alignSelf: 'flex-start',
    fontSize: 16,
    padding: 0
  },
  card: {
    flex: 1,
    padding: 0,
    width: '100%'
  },
  question: {
    width: '80%'
  },
  conditionCard: {
    flex: 1,
    width: '100%'
  },
  secondaryCondition: {
    fontSize: 16,
    fontWeight: 'bold',
    marginBottom: height * 0.01
  },
  topDiagnosis: {
    flex: 1,
    width: '100%'
  },
  btnText: {
    textAlign: 'center',
    color: 'white',
    fontSize: 14
  },
  btnTextHolder: {
    borderWidth: 1,
    borderColor: 'rgba(0,0,0,0.5)',
    marginTop: height * 0.01,
    width: '100%'
  },
  Btn: {
    padding: 10,
    backgroundColor: '#004EFF'
  },
  button: {
    alignSelf: 'center',
    marginBottom: height * 0.05,
    width: '80%'
  },
  buttonText: {
    fontWeight: 'bold'
  }
});
