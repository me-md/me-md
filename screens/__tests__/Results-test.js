import React from 'react';
import NavigationTestUtils from 'react-navigation/NavigationTestUtils';
import renderer from 'react-test-renderer';

import Results from '../../screens/Results/Results';

jest.mock('expo', () => ({
  AppLoading: 'AppLoading',
}));

describe('Results', () => {
  jest.useFakeTimers();

  beforeEach(() => {
    NavigationTestUtils.resetInternalState();
  });

  it.skip(`should match a snapshot`, () => {

    const mockNavigation = {
      state: {
        params: {
          userInfo: {

          },
          symptomFollowup: {

          },
          location: {
            lat: 121,
            long: 25
          },
          stateAbbreviation: 'CO'
        }
      }
    };

    const tree = renderer.create(<Results navigation={mockNavigation} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

});
