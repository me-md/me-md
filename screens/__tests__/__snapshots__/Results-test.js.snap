// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results should match a snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "width": "100%",
    }
  }
>
  <ViewManagerAdapter_ExpoLinearGradient
    proxiedProperties={
      Object {
        "colors": Array [
          4278210303,
          4287155441,
        ],
        "endPoint": undefined,
        "locations": undefined,
        "startPoint": undefined,
      }
    }
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 200.1,
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "alignSelf": "center",
          "borderWidth": 0,
          "height": 133.4,
          "marginBottom": 13.34,
          "marginTop": 60.03,
          "width": 133.4,
        }
      }
    />
  </ViewManagerAdapter_ExpoLinearGradient>
  <RCTScrollView
    collapsable={false}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    style={
      Object {
        "flex": 1,
        "marginBottom": 53.36,
        "marginTop": 26.68,
        "width": "100%",
      }
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
              "borderBottomWidth": 0.5,
              "borderColor": "#ccc",
              "borderLeftWidth": 0.5,
              "borderRadius": 2,
              "borderRightWidth": 0.5,
              "borderTopWidth": 0.5,
              "elevation": 3,
              "flexWrap": "nowrap",
              "marginBottom": 5,
              "marginLeft": 2,
              "marginRight": 2,
              "marginTop": 5,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.1,
              "shadowRadius": 1.5,
            },
            Object {
              "alignSelf": "center",
              "flex": 1,
              "shadowColor": "black",
              "shadowOffset": Object {
                "height": 5,
                "width": 5,
              },
              "shadowOpacity": 0.3,
              "shadowRadius": 4.65,
              "width": "95%",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "borderRadius": 2,
              "flexDirection": "row",
              "paddingBottom": 10,
              "paddingLeft": 15,
              "paddingRight": 15,
              "paddingTop": 10,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "alignSelf": "stretch",
                "flex": 1,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 16,
                  },
                  Object {
                    "fontSize": 16,
                    "fontWeight": "bold",
                    "marginBottom": 13.34,
                    "marginTop": 2.668,
                  },
                ]
              }
              uppercase={false}
            >
              undefined (undefined)
            </Text>
            <View
              animated={true}
              animationConfig={
                Object {
                  "bounciness": 0,
                }
              }
              animationType="spring"
              indeterminate={false}
              indeterminateAnimationDuration={1000}
              onLayout={[Function]}
              progress={0}
              style={
                Array [
                  Object {
                    "backgroundColor": undefined,
                    "borderColor": "rgba(0, 122, 255, 1)",
                    "borderRadius": 4,
                    "borderWidth": 1,
                    "overflow": "hidden",
                    "width": 200,
                  },
                  Object {
                    "marginBottom": 13.34,
                  },
                ]
              }
              useNativeDriver={false}
            >
              <View
                style={
                  Object {
                    "backgroundColor": "rgba(0, 122, 255, 1)",
                    "height": 6,
                    "transform": Array [
                      Object {
                        "translateX": -7.105427357601002e-15,
                      },
                      Object {
                        "translateX": -99,
                      },
                      Object {
                        "scaleX": 0.0001,
                      },
                    ],
                  }
                }
              />
            </View>
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                }
              }
              uppercase={false}
            >
              Probability: NaN%
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
              "borderBottomWidth": 0.5,
              "borderColor": "#ccc",
              "borderLeftWidth": 0.5,
              "borderRadius": 2,
              "borderRightWidth": 0.5,
              "borderTopWidth": 0.5,
              "elevation": 3,
              "flexWrap": "nowrap",
              "marginBottom": 5,
              "marginLeft": 2,
              "marginRight": 2,
              "marginTop": 5,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.1,
              "shadowRadius": 1.5,
            },
            Object {
              "alignSelf": "center",
              "flex": 1,
              "shadowColor": "black",
              "shadowOffset": Object {
                "height": 5,
                "width": 5,
              },
              "shadowOpacity": 0.3,
              "shadowRadius": 4.65,
              "width": "95%",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "borderRadius": 2,
              "flexDirection": "row",
              "paddingBottom": 10,
              "paddingLeft": 15,
              "paddingRight": 15,
              "paddingTop": 10,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "alignSelf": "stretch",
                "flex": 1,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 16,
                  },
                  Object {
                    "fontSize": 16,
                    "fontWeight": "bold",
                    "marginBottom": 13.34,
                    "marginTop": 2.668,
                  },
                ]
              }
              uppercase={false}
            >
              undefined (undefined)
            </Text>
            <View
              animated={true}
              animationConfig={
                Object {
                  "bounciness": 0,
                }
              }
              animationType="spring"
              indeterminate={false}
              indeterminateAnimationDuration={1000}
              onLayout={[Function]}
              progress={0}
              style={
                Array [
                  Object {
                    "backgroundColor": undefined,
                    "borderColor": "rgba(0, 122, 255, 1)",
                    "borderRadius": 4,
                    "borderWidth": 1,
                    "overflow": "hidden",
                    "width": 200,
                  },
                  Object {
                    "marginBottom": 13.34,
                  },
                ]
              }
              useNativeDriver={false}
            >
              <View
                style={
                  Object {
                    "backgroundColor": "rgba(0, 122, 255, 1)",
                    "height": 6,
                    "transform": Array [
                      Object {
                        "translateX": -7.105427357601002e-15,
                      },
                      Object {
                        "translateX": -99,
                      },
                      Object {
                        "scaleX": 0.0001,
                      },
                    ],
                  }
                }
              />
            </View>
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                }
              }
              uppercase={false}
            >
              Probability: NaN%
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#007aff",
        "borderBottomWidth": null,
        "borderColor": "#007aff",
        "borderLeftWidth": null,
        "borderRadius": 57,
        "borderRightWidth": null,
        "borderTopWidth": null,
        "borderWidth": undefined,
        "elevation": 2,
        "flexDirection": "row",
        "height": 93.38000000000001,
        "justifyContent": "space-around",
        "marginBottom": 26.68,
        "marginTop": -333.5,
        "opacity": 1,
        "paddingBottom": 6,
        "paddingTop": 6,
        "shadowColor": "black",
        "shadowOffset": Object {
          "height": 5,
          "width": 5,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 4.65,
        "width": 487.5,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "color": "#fff",
            "fontFamily": "System",
            "fontSize": 16.5,
            "marginLeft": 0,
            "marginRight": 0,
            "paddingLeft": 16,
            "paddingRight": 16,
          },
          Object {
            "fontSize": 33.35,
            "fontWeight": "bold",
          },
        ]
      }
      uppercase={false}
    >
      Find Doctors
    </Text>
    <Text />
  </View>
</View>
`;
